Add TensorRT packages

aidalgol opened the pull request:
###### Description of changes

Add derivations for TensorRT, a high-performance deep learning interface SDK from NVIDIA, which is at this point non-redistributable.  Uses different upstream tarballs depending on the current CUDA version, similar to the cudnn derivations.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

aidalgol committed:
tensorrt: init at 8.4.0.6

Add derivation for TensorRT 8, a high-performance deep learning interface SDK
from NVIDIA, which is at this point non-redistributable.  The current version
aldo requires CUDA 11, so this is left out of the cudaPackages_10* scopes.

SuperSandro2000 commented on pkgs/development/python-modules/tensorrt/default.nix:
```diff
@@ -0,0 +1,45 @@
+, tensorrt-bin
+, cudaPackages
+, cudatoolkit
+
+, pyVersion
```
```suggestion
```

SuperSandro2000 commented on pkgs/applications/science/math/tensorrt/default.nix:
```diff
@@ -0,0 +1,49 @@
+
+stdenv.mkDerivation {
+  pname = "cudatoolkit-${cudatoolkit.majorVersion}-tensorrt";
+  version = "8.4.0.6";
```
```suggestion
  version = "8.4.0.6";
  
  outputs = [ "out" "dev" ];
```

SuperSandro2000 commented on pkgs/applications/science/math/tensorrt/default.nix:
```diff
@@ -0,0 +1,49 @@
+    cudaPackages.cudnn
+  ];
+
+  sourceRoot = "TensorRT-8.4.0.6";
```
Don't hardcode version numbers

aidalgol replied:
Fixed.

SuperSandro2000 reviewed (changes requested):
Missing meta

aidalgol commented:
@SuperSandro2000 I appreciate the quick feedback, but this is nowhere ready for review.  I created this draft PR at the suggestion of @SomeoneSerge, after much back and forth in the Nix CUDA Maintainers Matrix room, to ease collaboration efforts on this package.

SomeoneSerge commented:
Good! Now let's expose the expression e.g. as an atttribute, e.g. `cudaPackages.tensorrt`
You'd need to write an extension for `cudaPackages` namespace.
It's going to be a function of the form `final: prev: { tensorrt = final.callPackage ... { }; }` where `final` an instance of `cudaPackages`.
You'll add it to this list: https://github.com/NixOS/nixpkgs/blob/473669d4603b281946ea0c8d6007e0a6c1900a49/pkgs/top-level/cuda-packages.nix#L68
For now, I think you could define it right there, next to [`cutensorExtension`](https://github.com/NixOS/nixpkgs/blob/473669d4603b281946ea0c8d6007e0a6c1900a49/pkgs/top-level/cuda-packages.nix#L22) (which you can use as a template, by the way)

Also, as Sandro suggests, it won't hurt to start bringing things into a shape right away: let's add a `meta` attribute that would declare the `unfree` license, limit the `platforms` to (I suppose) `x86_64-linux`, and provide a link to some changelog s.t. it's easier to track updates at least manually; let's move the `version` into a `let ... in` that we can reuse it when forming the filenames

SomeoneSerge commented on pkgs/development/python-modules/tensorrt/default.nix:
```diff
@@ -0,0 +1,45 @@
+  pname = "tensorrt";
+  version = "8.4.0.6";
+
+  src = # refet to same file as 'tensorrt' package
```
@aidalgol  As a baseline, you can access `cudaPackages.tensorrt.src` from here after you've exposed it

aidalgol replied:
Fixed.

SomeoneSerge commented on pkgs/development/python-modules/tensorrt/default.nix:
```diff
@@ -0,0 +1,45 @@
+  preUnpack = ''
+    mkdir -p dist
+    tar --strip-components=2 -xf "$src" --directory=dist \
+      "TensorRT-8.4.0.6/python/tensorrt-8.4.0.6-cp${pyVersion}-none-linux_x86_64.whl"
```
Meaning: `8.4.0.6` -> `${version}`

aidalgol replied:
This is now a parameter.

aidalgol commented:
@SomeoneSerge With my latest changes, you should now be able to reproduce the error I was describing on matrix (from before creating this draft PR) by running `NIXPKGS_ALLOW_UNFREE=1 nix-build -A python39Packages.tensorrt`.  I'm not sure whether the `checkPhase` I have written is appropriate for normal regression testing once we have this working, but it makes testing changes at this stage easier.

SuperSandro2000 commented:
If you want to collaborate on very WIP things I would suggest to work in a PR or branch in your fork.

SomeoneSerge commented:
@SuperSandro2000 thank you. It explicitly was my suggestion to @aidalgol that we move the discussion from matrix straight into a draft PR in nixpkgs that we can iterate faster. This is a messy non-redist package, which means there's some trivia about manually gluing pieces together to handle in the beginning

With that done, we'll mark the draft as "ready for review" and tag you

Cheers

SomeoneSerge commented on pkgs/development/python-modules/tensorrt/default.nix:
```diff
@@ -0,0 +1,58 @@
+  ];
+
+  propagatedBuildInputs = [
+    cudatoolkit
```
I assume these are for autopatchelf. Move to `buildInputs`.

`propagatedBuildInputs` are really only used for python packages and largely because they don't have headers where we could write the rpaths; we avoid `propagatedBuildInputs`, because one can soon  shoot one's feet with multiple instances of the same package, especially when they're as obscure and heavy as cuda/cudnn

aidalgol replied:
Moved to `buildInputs`.

SomeoneSerge commented on pkgs/development/python-modules/tensorrt/default.nix:
```diff
@@ -0,0 +1,58 @@
+  sourceRoot = ".";
+
+  runtimeDependencies = [
+    cudaPackages.tensorrt
```
IIRC, these are the paths that `autoPatchelf` adds to all binaries unconditionally (even when they're not "needed"). Do we want this behaviour?

SomeoneSerge commented on pkgs/development/python-modules/tensorrt/default.nix:
```diff
@@ -0,0 +1,58 @@
+
+  checkPhase = ''
+    # tensorrt.Builder may fail into an infinite loop, hence the timeout.
+    timeout 3 python3 -c 'import tensorrt; assert tensorrt.Builder(tensorrt.Logger())'
```
I see.

This gives `[06/22/2022-17:53:03] [TRT] [E] 1: [checkMacros.cpp::catchCudaError::271] Error Code 1: Cuda Runtime (CUDA driver version is insufficient for CUDA runtime version)`.

If you use the `cudaStub` expression from pytorch like so: `${cudaStubEnv} timeout 3 python3 -c 'import tensorrt; assert tensorrt.Builder(tensorrt.Logger())'`, you'll get a different error:
`[06/22/2022-18:00:52] [TRT] [E] 1: [checkMacros.cpp::catchCudaError::271] Error Code 1: Cuda Runtime (CUDA driver is a stub library)`

The former happens because trt cannot locate `libcuda.so`, the userspace cuda driver.The userspace driver in nixos is deployed in `/run/opengl-driver/lib`. This path is not available in nix sandbox, nor would the actual gpu be accessible either.

In the latter case we set up `LD_LIBRARY_PATH` to point at a directory with a fake `libcuda.so`, but it would seem one simply cannot construct `tensorrt.Builder` with that.

Unless there are some bits in trt that could be tested without a gpu, we'll have to limit ourselves to simply verifying that `tensorrt` is importable. By the way, this could be abbreviated as follows:

```nix
  pythonCheckImports = [
    "tensorrt"
  ];
```

You'll have to test the actual functionality manually, outside the sandbox

SomeoneSerge replied:
Oh, and we're missing `/run/opengl-driver/lib` in the rpath as well.
Use  `cudaPackages.autoAddOpenGLRunpathHook` (I hope I spelled it right, it's quite a name)

aidalgol replied:
> This gives `[06/22/2022-17:53:03] [TRT] [E] 1: [checkMacros.cpp::catchCudaError::271] Error Code 1: Cuda Runtime (CUDA driver version is insufficient for CUDA runtime version)`.
> […]
> You'll have to test the actual functionality manually, outside the sandbox

I was afraid of something like this.  I get the exact same error message outside of the sandbox, so I could not be sure.  Now I know.

SomeoneSerge commented on pkgs/development/libraries/science/math/tensorrt/generic.nix:
```diff
@@ -0,0 +1,66 @@
+, stdenv
+, callPackage
+, cudatoolkit
+, cudaPackages
```
```diff
- , cudaPackages
```

We have a convention that expressions inside `cudaPackages` scope ask directly for individual `cudaPackages` elements (like `cudatoolki`, `cuda_nvcc`, etc). In contrast, external expressions only ask for the entirety of `cudaPackages`

SomeoneSerge commented on pkgs/development/python-modules/tensorrt/default.nix:
```diff
@@ -0,0 +1,58 @@
+, autoPatchelfHook
+, unzip
+, cudaPackages
+, cudatoolkit
```
```diff
- , cudatoolkit
```

Cf. comment about `cudaPackages` above.
The convention is to `inherit (cudaPackages) cudatoolkit;` instead

aidalgol commented:
@SomeoneSerge With your suggested changes, I am back to where I was before moving this to a `nixpkgs` branch.
```console
$ NIX_PATH=nixpkgs=$PWD NIXPKGS_ALLOW_UNFREE=1 nix-shell -p 'python310.withPackages (pypkgs: [ pypkgs.tensorrt ])'
$ python3 -c 'import tensorrt; assert tensorrt.Builder(tensorrt.Logger())'
[06/23/2022-12:09:18] [TRT] [E] 6: [libLoader.h::DynamicLibrary::49] Error Code 6: Internal Error (Unable to load library: libnvinfer_builder_resource.so.8.4.0)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: pybind11::init(): factory function returned nullptr
```
From running the same `python3` command under `strace`, something appears to be looking for `libnvinfer_builder_resource.so.8.4.0` in the wrong places:
```console
openat(AT_FDCWD, "/run/opengl-driver/lib/libnvinfer_builder_resource.so.8.4.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/nix/store/wk3xam5pd0a0jwirggg5zpbjzw8zzaf3-gcc-11.3.0-lib/lib/libnvinfer_builder_resource.so.8.4.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/nix/store/fz33c1mfi2krpg1lwzizfw28kj705yg0-glibc-2.34-210/lib/libnvinfer_builder_resource.so.8.4.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
```
I have no idea _what_ is looking for it here, though.

samuela commented on pkgs/development/python-modules/tensorrt/default.nix:
```diff
@@ -33,20 +33,15 @@ buildPythonPackage rec {
+  buildInputs = [
+    cudaPackages.cudatoolkit
     cudaPackages.cudnn
     cudaPackages.tensorrt
```
What's the difference between `cudaPackages.tensorrt` and the `tensorrt` in `pkgs/development/libraries/science/math/tensorrt/default.nix`?

aidalgol replied:
It should be the same package.

samuela commented on pkgs/development/libraries/science/math/tensorrt/generic.nix:
```diff
@@ -25,16 +26,14 @@ stdenv.mkDerivation {
-  propagatedBuildInputs = [
     cudatoolkit
-    cudaPackages.cudnn
+    cudnn
```
Is there a reason not to use the cudaPackages version?

aidalgol replied:
If I understood @SomeoneSerge's feedback on my earlier commits, this is the convention that CUDA packages should follow.

samuela commented on pkgs/development/python-modules/tensorrt/default.nix:
```diff
@@ -33,20 +33,15 @@ buildPythonPackage rec {
-  propagatedBuildInputs = [
-    cudatoolkit
+  buildInputs = [
+    cudaPackages.cudatoolkit
```
Can we use the new individual cudaPackages instead? We're trying to move away from `cudaPackages.cudatoolkit` as much as possible.

aidalgol replied:
Looks as if we don't even need anything from `cudatoolkit` in this package.  I removed it and rebuilt the package to see what `autoPatchelf` failed to find, and it built just find, and I'm not getting a different error when I try to use the `tensorrt` python module.

samuela reviewed (commented):
Thanks for showing these packages some love @aidalgol! It's def a step forward.

I'm afraid I may not have enough context to be very useful here, since I don't use TensorRT personally. I just have a few notes re the cudaPackages migration that we've been working on.

SomeoneSerge commented on pkgs/development/libraries/science/math/tensorrt/generic.nix:
```diff
@@ -0,0 +1,65 @@
+  # runtime.
+  postFixup = ''
+    echo 'Patching RPATH of libnvinfer libs'
+    patchelf --debug --add-rpath '$ORIGIN' \
```
This seems to execute before `autoPatchelfHook`, s.t. autoPatchelf removes `$ORIGIN` for the final rpath.
You can verify that by building `cudaPackages.tensorrt` and checking out `patchelf --print-rpath ./result/lib/libnvinfer*`
The test you suggested before (`assert tensorrt.Builder(tensorrt.Logger())`) fails looking for `libnvinfer.*` things.

It would be nice to have a way to solve this without directly using `patchelf`, but a solution that'll work for now is to modify not the `rpath`, but the list of "needed" dependencies so that `autoPatchelf` knows which paths to keep:

```
  # Inform autoPatchelf to keep $ORIGIN
  postFixup = ''
    echo 'Patching NEEDED of libnvinfer libs'
    patchelf --debug --add-needed 'libnvinfer.so' \
      "$out/lib/libnvinfer.so.${mostOfVersion}" \
      "$out/lib/libnvinfer_plugin.so.${mostOfVersion}" \
      "$out/lib/libnvinfer_builder_resource.so.${mostOfVersion}"
  '';
```

aidalgol committed:
tensorrt: support multiple CUDA versions

Refactor derivation to pick the version that supports the current CUDA version.
Based on the implementation of the same concept in the cudnn derivation.

aidalgol committed:
cudnn: fix incorrect hash

Hash mismatch for cudnn_8_3_2 discovered when building
cudaPackages_10_2.tensorrt, which depends on this version of cudnn.

samuela reviewed (approved):
I'm not part of the NVIDIA dev program, so I don't have a means to test building this but the changes LGTM. 

Any remaining hold ups on merging this?

aidalgol commented:
> Any remaining hold ups on merging this?

Nothing on my end.  Just waiting for my last set of changes to be reviewed.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-install-a-specific-version-of-cuda-and-cudnn/21725/4
